persist/fs/read.go

```diff
@@ -120,22 +121,55 @@ func NewReader(
 	}, nil
 }
 
-func (r *reader) Open(namespace ident.ID, shard uint32, blockStart time.Time) error {
-	var err error
+func (r *reader) Open(opts ReaderOpenOptions) error {
+	var (
+		namespace     = opts.Namespace
+		shard         = opts.Shard
+		blockStart    = opts.BlockStart
+		snapshotIndex = opts.SnapshotIndex
+		err           error
+	)
+
+	var shardDir string
+	var checkpointFilePath string
+	switch opts.FilesetType {
+	case persist.FilesetSnapshotType:
+		shardDir = ShardSnapshotsDirPath(r.filePathPrefix, namespace, shard)
+		checkpointFilePath = snapshotPathFromTimeAndIndex(shardDir, blockStart, checkpointFileSuffix, snapshotIndex)
+	case persist.FilesetFlushType:
+		shardDir = ShardDataDirPath(r.filePathPrefix, namespace, shard)
+		checkpointFilePath = filesetPathFromTime(shardDir, blockStart, checkpointFileSuffix)
+	default:
+		return fmt.Errorf("unable to open reader with fileset type: %s", opts.FilesetType)
+	}
 
 	// If there is no checkpoint file, don't read the data files.
-	shardDir := ShardDirPath(r.filePathPrefix, namespace, shard)
-	if err := r.readCheckpointFile(shardDir, blockStart); err != nil {
+	if err := r.readCheckpointFile(checkpointFilePath, blockStart); err != nil {
 		return err
 	}
 
 	var infoFd, digestFd *os.File
-	if err := openFiles(os.Open, map[string]**os.File{
-		filesetPathFromTime(shardDir, blockStart, infoFileSuffix):        &infoFd,
-		filesetPathFromTime(shardDir, blockStart, digestFileSuffix):      &digestFd,
-		filesetPathFromTime(shardDir, blockStart, bloomFilterFileSuffix): &r.bloomFilterFd,
-	}); err != nil {
-		return err
+	switch opts.FilesetType {
+	case persist.FilesetSnapshotType:
```

Comment:
could you do the type -> filename resolution in a helper? mainly want to avoid the duplication of `openFiles` call.

Reply:
Didn't end up doing this in a helper, but cleaned it up a lot